File: server/roster.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { LIVE_MS } from "../lib/config.ts";
import { newestById, pidAlive, readJsonFile } from "../lib/node.ts";
import { pathUnder } from "../lib/under.ts";
import type { OrcaFocus, OrcaPane, SessionRow } from "../lib/types.ts";
import { readClaudeSessions } from "./claude.ts";
import { readCodexSessions } from "./codex.ts";
import { readOrcaFocus, readOrcaLivePanes } from "./orca.ts";

export function grokHome(): string {
  return process.env.GROK_HOME || path.join(os.homedir(), ".grok");
}

export function encodeCwd(cwd: string): string {
  return encodeURIComponent(cwd);
}

export function sessionDir(cwd: string, sessionId: string, home = grokHome()): string {
  const direct = path.join(home, "sessions", encodeCwd(cwd), sessionId);
  if (fs.existsSync(direct)) return direct;
  const root = path.join(home, "sessions");
  if (!fs.existsSync(root)) return direct;
  let groups: fs.Dirent[];
  try {
    groups = fs.readdirSync(root, { withFileTypes: true });
  } catch {
    return direct;
  }
  for (const group of groups) {
    if (!group.isDirectory()) continue;
    const candidate = path.join(root, group.name, sessionId);
    if (fs.existsSync(candidate)) return candidate;
  }
  return direct;
}

export function updatesPath(session: SessionRow, home = grokHome()): string {
  if (session?.updates) return session.updates;
  const id = session.nativeId || session.session_id;
  return path.join(sessionDir(session.cwd, id, home), "updates.jsonl");
}

type SessionSummary = {
  generated_title?: string;
  agent_name?: string;
};

export function readSummary(session: SessionRow, home = grokHome()): SessionSummary | null {
  const file = path.join(sessionDir(session.cwd, session.session_id, home), "summary.json");
  return readJsonFile<SessionSummary | null>(file, null);
}

function findSessionCwd(sessionId: string, home: string): string | null {
  const root = path.join(home, "sessions");
  if (!fs.existsSync(root)) return null;
  let groups: fs.Dirent[];
  try {
    groups = fs.readdirSync(root, { withFileTypes: true });
  } catch {
    return null;
  }
  for (const group of groups) {
    if (!group.isDirectory()) continue;
    if (!fs.existsSync(path.join(root, group.name, sessionId))) continue;
    try {
      return decodeURIComponent(group.name);
    } catch {
      return group.name;
    }
  }
  return null;
}

function rosterEntry(row: SessionRow, home: string, { live = false } = {}): SessionRow {
  const summary = readSummary(row, home);
  const updates = updatesPath(row, home);
  let mtime = 0;
  try {
    mtime = fs.statSync(updates).mtimeMs;
  } catch {
    mtime = 0;
  }
  return {
    session_id: row.session_id,
    pid: row.pid || 0,
    cwd: row.cwd,
    opened_at: row.opened_at,
    title: summary?.generated_title || path.basename(row.cwd),
    agent: summary?.agent_name || "agent",
    provider: row.provider || "grok",
    updates,
    mtime,
    live,
  };
}

function readActiveSessionRows(home: string): SessionRow[] {
  const rows = readJsonFile<unknown>(path.join(home, "active_sessions.json"), []);
  return Array.isArray(rows)
    ? rows.filter((row): row is SessionRow => Boolean(row?.session_id && row.cwd))
    : [];
}

export function readActiveSessions(home = grokHome(), panes?: Map<string, OrcaPane> | null): SessionRow[] {
  const rows = readActiveSessionRows(home);
  const byId = new Map<string, SessionRow>();
  for (const row of rows) {
    const hostLive = pidAlive(row.pid);
    let mtime = 0;
    try {
      mtime = fs.statSync(updatesPath(row, home)).mtimeMs;
    } catch {
      mtime = 0;
    }
    const recent = Date.now() - mtime < LIVE_MS;
    byId.set(row.session_id, rosterEntry(row, home, { live: hostLive || recent }));
  }

  const livePanes = panes === undefined && home === grokHome() ? readOrcaLivePanes() : panes;
  const out: SessionRow[] = [];
  const seen = new Set<string>();
  if (livePanes) {
    for (const [id, pane] of livePanes) {
      const existing = byId.get(id);
      if (existing) {
        existing.live = true;
        out.push(existing);
        seen.add(id);
        continue;
      }
      const cwd = pane?.cwd || findSessionCwd(id, home);
      if (!cwd) continue;
      out.push(
        rosterEntry(
          { session_id: id, cwd, pid: 0, provider: "grok", title: "", agent: "agent", mtime: 0, live: true },
          home,
          { live: true },
        ),
      );
      seen.add(id);
    }
  }
  for (const session of byId.values()) {
    if (seen.has(session.session_id)) continue;
    if (!session.live) continue;
    out.push(session);
    seen.add(session.session_id);
  }
  out.sort((a, b) => (b.mtime || 0) - (a.mtime || 0));
  return out;
}

export function readAllSessions(home = grokHome()): SessionRow[] {
  return newestById([...readActiveSessions(home), ...readClaudeSessions(), ...readCodexSessions()]);
}

export function rosterFingerprint(roster: SessionRow[] | null | undefined): string {
  return (roster || [])
    .map((s) => `${s.session_id}:${s.pid || 0}:${s.cwd}:${s.title}:${s.live ? 1 : 0}`)
    .join("|");
}

export type FocusHint = OrcaFocus | string | null | undefined;

function focusHint(hint: FocusHint): OrcaFocus {
  if (hint && typeof hint === "object") return hint;
  if (typeof hint === "string" && hint) return { cwd: hint, sessionId: null };
  return readOrcaFocus() || { cwd: null, sessionId: null };
}

export function pickFocusedSession(
  roster: SessionRow[] | null | undefined,
  hint: FocusHint = readOrcaFocus(),
): SessionRow | null {
  if (!roster?.length) return null;
  const focus = focusHint(hint);
  if (focus.sessionId) {
    const byId = roster.find((s) => s.session_id === focus.sessionId);
    if (byId) return byId;
  }
  const cwd = focus.cwd;
  if (cwd) {
    const exact = roster.filter((s) => s.cwd === cwd);
    const nested = roster.filter((s) => pathUnder(cwd, s.cwd));
    const pool = exact.length ? exact : nested;
    if (pool.length) {
      return pool.slice().sort((a, b) => Number(b.live) - Number(a.live) || (b.mtime || 0) - (a.mtime || 0))[0];
    }
  }
  return null;
}

File: internal/roster/roster.go
package roster

import (
	"encoding/json"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/ryusudol/plexus/internal/claude"
	"github.com/ryusudol/plexus/internal/codex"
	"github.com/ryusudol/plexus/internal/jsonx"
	"github.com/ryusudol/plexus/internal/orca"
	"github.com/ryusudol/plexus/internal/paths"
	"github.com/ryusudol/plexus/internal/types"
	"github.com/ryusudol/plexus/internal/under"
)

func GrokHome() string {
	if env := os.Getenv("GROK_HOME"); env != "" {
		return env
	}
	return filepath.Join(paths.Home(), ".grok")
}

func EncodeCwd(cwd string) string {
	return url.QueryEscape(cwd)
}

func SessionDir(cwd, sessionID, home string) string {
	if home == "" {
		home = GrokHome()
	}
	direct := filepath.Join(home, "sessions", EncodeCwd(cwd), sessionID)
	if _, err := os.Stat(direct); err == nil {
		return direct
	}
	root := filepath.Join(home, "sessions")
	entries, err := os.ReadDir(root)
	if err != nil {
		return direct
	}
	for _, group := range entries {
		if !group.IsDir() {
			continue
		}
		candidate := filepath.Join(root, group.Name(), sessionID)
		if _, err := os.Stat(candidate); err == nil {
			return candidate
		}
	}
	return direct
}

func UpdatesPath(session types.SessionRow, home string) string {
	if session.Updates != "" {
		return session.Updates
	}
	if home == "" {
		home = GrokHome()
	}
	id := session.NativeID
	if id == "" {
		id = session.SessionID
	}
	return filepath.Join(SessionDir(session.Cwd, id, home), "updates.jsonl")
}

type summary struct {
	GeneratedTitle string `json:"generated_title"`
	AgentName      string `json:"agent_name"`
}

func readSummary(session types.SessionRow, home string) *summary {
	file := filepath.Join(SessionDir(session.Cwd, session.SessionID, home), "summary.json")
	b, err := os.ReadFile(file)
	if err != nil {
		return nil
	}
	var s summary
	if json.Unmarshal(b, &s) != nil {
		return nil
	}
	return &s
}

func findSessionCwd(sessionID, home string) string {
	root := filepath.Join(home, "sessions")
	entries, err := os.ReadDir(root)
	if err != nil {
		return ""
	}
	for _, group := range entries {
		if !group.IsDir() {
			continue
		}
		if _, err := os.Stat(filepath.Join(root, group.Name(), sessionID)); err != nil {
			continue
		}
		dec, err := url.QueryUnescape(group.Name())
		if err != nil {
			return group.Name()
		}
		return dec
	}
	return ""
}

func rosterEntry(row types.SessionRow, home string, live bool) types.SessionRow {
	sum := readSummary(row, home)
	updates := UpdatesPath(row, home)
	var mtime int64
	if st, err := os.Stat(updates); err == nil {
		mtime = st.ModTime().UnixMilli()
	}
	title := filepath.Base(row.Cwd)
	agent := "agent"
	if sum != nil {
		if sum.GeneratedTitle != "" {
			title = sum.GeneratedTitle
		}
		if sum.AgentName != "" {
			agent = sum.AgentName
		}
	}
	provider := row.Provider
	if provider == "" {
		provider = "grok"
	}
	return types.SessionRow{
		SessionID: row.SessionID,
		PID:       row.PID,
		Cwd:       row.Cwd,
		OpenedAt:  row.OpenedAt,
		Title:     title,
		Agent:     agent,
		Provider:  provider,
		Updates:   updates,
		Mtime:     mtime,
		Live:      live,
	}
}

func readActiveSessionRows(home string) []types.SessionRow {
	b, err := os.ReadFile(filepath.Join(home, "active_sessions.json"))
	if err != nil {
		return nil
	}
	var raw []map[string]any
	if json.Unmarshal(b, &raw) != nil {
		return nil
	}
	out := []types.SessionRow{}
	for _, rec := range raw {
		id := jsonx.Str(rec["session_id"])
		cwd := jsonx.Str(rec["cwd"])
		if id == "" || cwd == "" {
			continue
		}
		out = append(out, types.SessionRow{
			SessionID: id,
			PID:       jsonx.Int(rec["pid"]),
			Cwd:       cwd,
			OpenedAt:  jsonx.Str(rec["opened_at"]),
		})
	}
	return out
}

func ReadActiveSessions(home string, panes map[string]orca.Pane, panesSet bool) []types.SessionRow {
	if home == "" {
		home = GrokHome()
	}
	activePath := filepath.Join(home, "active_sessions.json")
	_, activeErr := os.Stat(activePath)
	hasActiveFile := activeErr == nil
	rows := readActiveSessionRows(home)

	var livePanes map[string]orca.Pane
	var hasPanes bool
	if panesSet {
		livePanes, hasPanes = panes, true
	} else if home == GrokHome() {
		livePanes, hasPanes = orca.ReadLivePanes("")
	}

	pidCount := map[int]int{}
	for _, row := range rows {
		if row.PID > 0 {
			pidCount[row.PID]++
		}
	}

	byID := map[string]types.SessionRow{}
	for _, row := range rows {
		inPane := hasPanes && livePanes != nil && paneListed(livePanes, row.SessionID)
		hostLive := paths.PidAlive(row.PID)
		if !inPane && !hostLive {
			continue
		}
		if !inPane && hostLive && pidCount[row.PID] > 1 && siblingOnPane(rows, livePanes, hasPanes, row) {
			continue
		}
		byID[row.SessionID] = rosterEntry(row, home, true)
	}

	out := []types.SessionRow{}
	seen := map[string]bool{}
	if hasPanes && livePanes != nil {
		for id, pane := range livePanes {
			if existing, ok := byID[id]; ok {
				existing.Live = true
				out = append(out, existing)
				seen[id] = true
				continue
			}
			// Grok's active_sessions.json is authoritative once it exists.
			// A leftover Orca pane after terminate must not keep the session
			// in the picker.
			if hasActiveFile {
				continue
			}
			cwd := pane.Cwd
			if cwd == "" {
				cwd = findSessionCwd(id, home)
			}
			if cwd == "" {
				continue
			}
			out = append(out, rosterEntry(types.SessionRow{
				SessionID: id, Cwd: cwd, PID: 0, Provider: "grok", Title: "", Agent: "agent", Mtime: 0, Live: true,
			}, home, true))
			seen[id] = true
		}
	}
	for _, session := range byID {
		if seen[session.SessionID] {
			continue
		}
		out = append(out, session)
		seen[session.SessionID] = true
	}
	return types.NewestByID(out)
}

func paneListed(panes map[string]orca.Pane, id string) bool {
	if panes == nil {
		return false
	}
	_, ok := panes[id]
	return ok
}

func siblingOnPane(rows []types.SessionRow, panes map[string]orca.Pane, hasPanes bool, row types.SessionRow) bool {
	if !hasPanes || panes == nil || row.PID <= 0 {
		return false
	}
	for _, other := range rows {
		if other.SessionID == row.SessionID || other.PID != row.PID {
			continue
		}
		if paneListed(panes, other.SessionID) {
			return true
		}
	}
	return false
}

func ReadAllSessions(home string) []types.SessionRow {
	if home == "" {
		home = GrokHome()
	}
	all := append([]types.SessionRow{}, ReadActiveSessions(home, nil, false)...)
	all = append(all, claude.ReadSessions("", time.Time{})...)
	all = append(all, codex.ReadSessions("", time.Time{})...)
	return types.NewestByID(all)
}

func Fingerprint(roster []types.SessionRow) string {
	parts := make([]string, 0, len(roster))
	for _, s := range roster {
		live := "0"
		if s.Live {
			live = "1"
		}
		parts = append(parts, s.SessionID+":"+strconv.Itoa(s.PID)+":"+s.Cwd+":"+s.Title+":"+live)
	}
	return strings.Join(parts, "|")
}

func PickFocusedSession(list []types.SessionRow, hint orca.Focus) *types.SessionRow {
	if len(list) == 0 {
		return nil
	}
	if hint.SessionID != "" {
		for i := range list {
			if list[i].SessionID == hint.SessionID {
				return &list[i]
			}
		}
	}
	cwd := hint.Cwd
	if cwd == "" {
		return nil
	}
	exact := []types.SessionRow{}
	nested := []types.SessionRow{}
	for _, s := range list {
		if s.Cwd == cwd {
			exact = append(exact, s)
		}
		if under.PathUnder(cwd, s.Cwd) {
			nested = append(nested, s)
		}
	}
	pool := exact
	if len(pool) == 0 {
		pool = nested
	}
	if len(pool) == 0 {
		return nil
	}
	best := pool[0]
	for _, s := range pool[1:] {
		if boolInt(s.Live) > boolInt(best.Live) || (boolInt(s.Live) == boolInt(best.Live) && s.Mtime > best.Mtime) {
			best = s
		}
	}
	return &best
}

func boolInt(v bool) int {
	if v {
		return 1
	}
	return 0
}
